include/core/CProgramCounters.h

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+#ifndef INCLUDED_ml_core_CProgramCounters_h
+#define INCLUDED_ml_core_CProgramCounters_h
+
+#include <core/CNonCopyable.h>
+#include <core/ImportExport.h>
+
+#include <array>
+#include <iosfwd>
+#include <set>
+#include <string>
+#include <vector>
+
+#include <stdint.h>
+
+class CProgramCountersTest;
+
+namespace ml {
+namespace counter_t {
+
+//! Changing the order of these enumeration values will corrupt persisted model
+//! state, so don't.  Any new counters should be added in the penultimate
+//! position in the enum, immediately before E_LastEnumCounter.
```

Comment:
this can be avoided by explicitly assigning values to the enum values, especially if the enum is used for persistence/communication

Also think about using a scoped enum.

Reply:
The only slight downside of this is it forces us to have a much larger array at the start since we use direct addressing based on the enum value. I think we have to keep this so the lookup remains cheap.

Reply:
+1,

An argument against reserving ranges is the use of an array for the counters, so we would space for unused counters between the ranges.

So simple increment could work and as we explicitly set the value, you can group the code the way you like it. The highest value could be a constant outside of the enum. Programming errors like forgetting to update this constant for the highest enum, double assigning the same value, could be mitigated with a unit test which loops over the definitions. Maybe thats even an argument to not allow any holes.

Reply:
True, that makes two downsides to leaving gaps then, as validation complexity is another problem then.  So maybe we should stick with the way it's done now.  We just need a policy on whether the names are ordered numerically or grouped by the process they go with.  For example, if we add a new `TSAD` counter after we've added the counters for "new analytics", should it go next to the existing `TSAD` names or at the end of the list?

Either way, I think it's safest to assign every single name an explicit value.